lib/post_request_task/task.py

```diff
@@ -0,0 +1,113 @@
+import threading
+from functools import partial
+
+from django.core.signals import (got_request_exception, request_finished,
+                                 request_started)
+
+import commonware.log
+from celery import task as base_task
+from celery import Task
+
+
+log = commonware.log.getLogger('z.post_request_task')
+
+
+_locals = threading.local()
+
+
+def _get_task_queue():
+    """Returns the calling thread's task queue."""
+    return _locals.__dict__.setdefault('task_queue', [])
+
+
+def _start_queuing_tasks(**kwargs):
+    """Starts queuing tasks for this thread.
+
+    Not supposed to be called directly, instead it's connected to the
+    request_started signal.
+
+    If not called, tasks are delayed normally (so tasks still function without
+    having to call _send_tasks_and_stop_queuing() manually when we're outside
+    the request-response cycle.."""
+    _locals.__dict__['queue_tasks'] = True
+
+
+def _stop_queuing_tasks(**kwargs):
+    """Stops queuing tasks for this thread.
+
+    Not supposed to be called directly, only useful for tests cleanup."""
+    _locals.__dict__.pop('queue_tasks', None)
+
+
+def is_task_queuing_enabled_for_this_thread():
+    """Returns whether post request task queuing is enabled for this thread."""
+    return _locals.__dict__.get('queue_tasks', False)
+
+
+def _send_tasks_and_stop_queuing(**kwargs):
+    """Sends all delayed Celery tasks and stop queuing new ones for now."""
+    _locals.__dict__.pop('queue_tasks', None)
+    queue = _get_task_queue()
+    while queue:
+        cls, args, kwargs = queue.pop(0)
+        cls.original_apply_async(*args, **kwargs)
+
+
+def _discard_tasks(**kwargs):
+    """Discards all delayed Celery tasks."""
```

Comment:
I would like to see logging here so we know which tasks were discarded in an exception.